### ISSUE TITLE
odo project set doesn't work

### ISSUE THREAD

**Issue by deboer-tim** [openshift/odo#4118]
/kind bug

## What versions of software are you using?

**Operating System:**
MacOS Catalina

**Output of `odo version`:**
odo v2.0.0 (6fbb9d9bf)

Server: https://api.member.crt-stage.com:6443
Kubernetes: v1.18.3+5302882

## How did you run odo exactly?
I created a project 'timproj' and successfully pushed a component to a cluster. Many days later, I connected to another cluster and tried to push the same component. It failed and I couldn't find any way to see that I was still using timproj, nor to change it.

## Actual behavior
`deboer-mac:node2 deboer$ odo project list
ACTIVE     NAME
*          deboerrh-code
           deboerrh-dev
           deboerrh-stage
deboer-mac:node2 deboer$ odo push
 ✗  Failed to start component with name node2. Error: Failed to create the component: unable to determine if component node2 exists: deployments.apps "node2" is forbidden: User "deboerrh" cannot get resource "deployments" in API group "apps" in the namespace "timproj"
deboer-mac:node2 deboer$ odo project set deboerrh-code
Already on project : deboerrh-code
deboer-mac:node2 deboer$ odo push
 ✗  Failed to start component with name node2. Error: Failed to create the component: unable to determine if component node2 exists: deployments.apps "node2" is forbidden: User "deboerrh" cannot get resource "deployments" in API group "apps" in the namespace "timproj"`

I was not able to push the component successfully until I hand-edited .odo/env/env.yaml and fixed the ComponentSettings/Project.

## Expected behavior
'odo project list' should have shown I was still using 'timproj' (likely with an indication that it doesn't exist on the cluster), and 'odo project set' should have fixed it.

**Comment by kadel** [comment]
Project information is saved in the time of component creation (`odo create`).
This is to ensure that you are using the correct project when you are switching between multiple odo components in different projects. 

To change the project for the existing component you need to do `odo env set project  deboerrh-code`. (Note that changing project for existing odo components that has been push can have unintended consequences - It will re-create the component in a new project)

`odo project set` command is intended for switching project in kubeconfig but it doesn't affect existing the odo component.


I know that this is a little bit confusing and we need to better document this.

**Comment by deboer-tim** [comment]
When the user creates the project they have no understanding that it's actually setting two separate things, and in the typical flow they'd never see that there are env settings. Based on the interaction above it just looks like `odo project set` is failing, so users wouldn't even look at docs or for other solutions. The fact that they're both called project is even more confusing. IMHO this needs changes to the cli and/or messages, not just documentation.

**Comment by kadel** [comment]
> The fact that they're both called project is even more confusing. IMHO this needs changes to the cli and/or messages, not just documentation.

Do you have any suggestions on how to make this a better experience?

The current flow is:
 1) choose a project to work with one of the following steps
   a) `odo project create myproject` - to create a new project
   b) `odo project set myproject`  - to switch to a and existing project
   c) do nothing to use the current project
2) `odo create` to create new odo component in a selected project
3) if you need to move the existing odo component to a different project you should make sure that it is deleted from the cluster (`odo delete`), than change project (`odo env set project`), and push again (`odo push`).

As I mentioned before changing the project for the pushed components has a lot of issues that would be quite hard to solve (you can't easily move PVC between namespaces, odo would have to handle data transfer itself).
I would consider removing the option for changing the project for odo components. In case someone needs this they will have to re-create component (which is essentially what they need to do anyway). And make clear that `odo project` changes currently active project/namespace in kubeconfig not for the component. 
Don't know what else we could do.

**Comment by girishramnani** [comment]
I dont think developers care about their storage when in development mode, so if someone is changing the project or are coming to the component after a while they just want the component to run.
So I would suggest that we try to create the project if its not found because in most cases thats what the user also expects.

**Comment by deboer-tim** [comment]
Agreed, as a user I don't really care about storage or what was there before, I just want to run the current thing. If the project is wrong or invalid I'm ok with having to create/fix before push, but I kind of expect any required cleanup (e.g. delete) to happen automatically. In other words, push takes what I have and syncs it to the cluster, not merges what I have. (and yes, I know there are edge cases where maybe I need to clean up)

Thinking along these lines, I still get back to the same two things in the description that I think would improve this:

1. If a user goes through the normal flow they've only used `odo project` commands to create or set the project, and under the covers this essentially did an `odo env set project`. If the user ever needs to change or fix the project, they're always going to default to trying `odo project` again. I think we should always do an `odo env set project` under the covers, not just the first time.

2. In the normal path when I create a project and then run `odo project list` I can see it. In a case like above it's confusing that the project I'm using isn't on the list, and the command is telling me a different project is active. In this case I really think 'odo project list' needs to include the project that the component is using, even if it's marked as invalid.
